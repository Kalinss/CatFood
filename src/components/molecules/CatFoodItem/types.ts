import React from "react";

export type CatFoodItemDataType = {
  id: number | string;
  description: {
    value: string;
    alternative: string;
  };
  title: string;
  subTitle: string;
  count: {
    value: string;
    metric: string;
  };
  info: string[];
  isActive: boolean;
  disabled: {
    value: boolean;
    warn: string;
  };
  additional: {
    value: string;
    alternative: string;
  };
  fullDescription:string
};
export type CatFoodItemEventClickType = (
  e: React.SyntheticEvent,
  id: string | number
) => void;
