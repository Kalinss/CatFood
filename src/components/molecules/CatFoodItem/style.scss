@import "../../../styles/variables";
@import "../../../styles/other";
.wrapper {
  display: flex;
  justify-content: flex-start;
  flex-direction: column;
}

.item {
  width: 320px;
  height: 480px;
  border-radius: 12px;
  border-top-left-radius: 100px;
  position: relative;
  cursor: pointer;
}

.border {
  position: absolute;
  z-index: 1;
  width: 100%;
  height: 100%;
}

.additional {
  color: white;
  font-size: 13px;
  text-align: center;
  margin-top: 15px;
  height: 13px;

  & button {
    color: $primary-decoration-color-default;
    outline: none;
    border: none;
    background-color: transparent;
    padding: 0;
    height: 100%;
    cursor: pointer;
  }
}

.main {
  position: relative;
  z-index: 2;
  height: 100%;
  padding-left: 45px;
  width: 100%;
  display: flex;
  flex-direction: column;
  padding-top: 20px;

  &:before {
    display: block;
    width: 259px;
    height: 269px;
    content: "";
    position: absolute;
    bottom: 4px;
    left: 4px;
    border-bottom-left-radius: 9px;
    background-image: url("../../../assets/images/cat.png");
    background-repeat: no-repeat;
  }
}

.subTitle {
  margin-top: 4px;
}

.warn {
  color: #ffff66;
}

.info {
  margin-top: 16px;
  font-size: 14px;
  color: $text-color-primary;

  & li {
    color: inherit;

    & b {
      color: inherit;
      font-weight: bold;
    }
  }

  & li:not(:first-child) {
    margin-top: 2px;
  }
}

.title {
  margin-top: 15px;
}

.price {
  position: absolute;
  bottom: 15px;
  right: 15px;
}

.description {
  font-size: 16px;
  color: $text-color-primary;
}

.item {
  & .price {
    transition: 0.4s ease;
    background-color: $primary-decoration-color-default;
  }

  & .border {
    transition: 0.4s ease;
    color: $primary-decoration-color-default;
  }
}

.item:hover {
  & .price {
    background-color: $primary-decoration-color-hover;
  }

  & .border {
    color: $primary-decoration-color-hover;
  }
}

.active {
  & .price {
    background-color: $second-decoration-color-default;
  }

  & .border {
    color: $second-decoration-color-default;
  }
}

.active:hover {
  & .price {
    background-color: $second-decoration-color-hover;
  }

  & .border {
    color: $second-decoration-color-hover;
  }
}

.disabled {
  cursor: default;
  background-color: #f2f2f2;
  opacity: 0.9;
  transition: 0.5s ease;

  & * {
    color: #d4d4d4;
  }

  & .price {
    & span {
      color: white;
    }
  }

  & .main {
    &:before {
      transition: 0.5s ease;
      filter: grayscale(50%);
      opacity: 0.6;
    }
  }

  & .price {
    background-color: #b3b3b3;
  }

  & .border {
    color: #b3b3b3;
  }
}

.disabled:hover {
  & .price {
    background-color: #b3b3b3;
  }

  & .border {
    color: #b3b3b3;
  }
}

.mouseLeave {
  animation: enter 0.4s ease;
}

.mouseEnter {
  animation: leave 0.4s ease;
}

@keyframes enter {
  0% {
    opacity: 1;
  }
  50% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
}

@keyframes leave {
  0% {
    opacity: 1;
  }
  50% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
}
