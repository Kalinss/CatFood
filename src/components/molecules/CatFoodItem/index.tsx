import React, { useState } from "react";
import style from "./style.scss";
import { BorderFoodItem } from "../../atoms/BorderFoodItem";
import { Header } from "../../atoms/Header";
import { CirclePriceLabel } from "../../atoms/CirclePriceLabel";
import classNames from "classnames";
import { getReactNode } from "../../../utils/other";
import { CatFoodItemDataType, CatFoodItemEventClickType } from "./types";

export type CatFoodItemType = {
  item: CatFoodItemDataType;
  clickHandler: CatFoodItemEventClickType;
  classStyle?: string;
};

export const CatFoodItem: React.FC<CatFoodItemType> = ({
  item,
  clickHandler = () => {},
  classStyle = "",
}) => {
  const [actuallyDescription, setActuallyDescription] = useState(
    item.description.value
  );
  const [isMouseEnter, setMouseEnter] = useState(false);
  const [protectFirstAnimation, setProtect] = useState(true);

  const getActuallyClass = (active: boolean, disabled: boolean) => {
    return classNames(
      style.item,
      active && style.active,
      disabled && style.disabled
    );
  };

  const additionalContent = (isDisabled: boolean, isActive: boolean) => {
    if (isDisabled) return getReactNode(item.disabled.warn);
    if (isActive) return getReactNode(item.additional.alternative);
    return getReactNode(item.additional.value);
  };

  // sideEffects, animation description text controller
  const ANIMATION_DURATION = 200; //ms, hideShow animation

  const toggleMouseAction = () => setMouseEnter(!isMouseEnter);

  const getDescriptionAnimationClass = () =>
    isMouseEnter ? style.mouseEnter : style.mouseLeave;

  const itemClickHandler = (e: React.SyntheticEvent, id: number | string) => {
    setActuallyDescription(item.description.value);
    setProtect(true);
    clickHandler(e, id);
  };

  const linkClickHandler = (e: React.SyntheticEvent, id: number | string) => {
    if ((e.target as HTMLElement).closest("a") === null) return;
    e.preventDefault();
    setActuallyDescription(item.description.value);
    setProtect(true);
    clickHandler(e, id);
  };

  const mouseEnterHandler = () => {
    if (!item.isActive) return;
    setProtect(false);
    toggleMouseAction();
    setTimeout(
      () => setActuallyDescription(item.description.alternative),
      ANIMATION_DURATION
    );
  };

  const mouseLeaveHandler = () => {
    if (!item.isActive || protectFirstAnimation) return;
    toggleMouseAction();
    setTimeout(
      () => setActuallyDescription(item.description.value),
      ANIMATION_DURATION
    );
  };

  return (
    <div className={classNames(style.wrapper, classStyle)}>
      <div
        className={getActuallyClass(item.isActive, item.disabled.value)}
        onClick={(e) => itemClickHandler(e, item.id)}
        onMouseEnter={mouseEnterHandler}
        onMouseLeave={mouseLeaveHandler}
      >
        <BorderFoodItem classStyle={style.border} />
        <div className={style.main}>
          <p
            className={classNames(
              style.description,
              getDescriptionAnimationClass()
            )}
          >
            {actuallyDescription}
          </p>
          <Header classStyle={style.title} as="h2">
            {item.title}
          </Header>
          <Header classStyle={style.subTitle}>{item.subTitle}</Header>
          <div className={style.info}>
            <ul>
              {item.info.map((value, i) => (
                <li key={i}>{getReactNode(value)}</li>
              ))}
            </ul>
          </div>
          <CirclePriceLabel
            classStyle={style.price}
            count={item.count.value}
            metric={item.count.metric}
          />
        </div>
      </div>
      <p
        onClick={(e) => linkClickHandler(e, item.id)}
        className={
          item.disabled.value
            ? classNames(style.additional, style.warn)
            : style.additional
        }
      >
        {additionalContent(item.disabled.value, item.isActive)}
      </p>
    </div>
  );
};
