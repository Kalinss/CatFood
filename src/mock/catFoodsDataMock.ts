import { CatFoodItemDataType } from "../components/molecules/CatFoodItem/types";

export const catFoodData: CatFoodItemDataType[] = [
  {
    id: 0,
    description: {
      value: "Сказочное заморское яство",
      alternative: "Котэ не одобряет?",
    },
    title: "Нямушка",
    subTitle: "c фуа-гра",
    count: {
      value: "0,5",
      metric: "кг",
    },
    info: ["<b>10</b> порций", "мышь в подарок"],
    isActive: false,
    disabled: {
      value: false,
      warn: "Печалька, с фуа-гра закончился.",
    },
    additional: {
      value: "Чего сидишь? Порадуй котэ, <button>купи</button>",
      alternative: "Печень утки разварная с артишоками.",
    },
    fullDescription:'Кошачий корм Нямушка 0.5 киллограмма, это 10 порций, мышь в подарок'
  },
  {
    id: 1,
    description: {
      value: "Сказочное заморское яство",
      alternative: "Котэ не одобряет?",
    },
    title: "Нямушка",
    subTitle: "c рыбой",
    count: {
      value: "2",
      metric: "кг",
    },
    info: ["<b>40</b> порций", "<b>2</b> мыши в подарок"],
    isActive: false,
    disabled: {
      value: false,
      warn: "Печалька, с рыбой закончился.",
    },
    additional: {
      value: "Чего сидишь? Порадуй котэ, <button>купи</button>",
      alternative: "Головы щучьи с чесноком да свежайшая сёмгушка.",
    },
    fullDescription:'Кошачий корм Нямушка 2 киллограмма, это 40 порций, 2 мыши в подарок'
  },
  {
    id: 2,
    description: {
      value: "Сказочное заморское яство",
      alternative: "Котэ не одобряет?",
    },
    title: "Нямушка",
    subTitle: "c курой",
    count: {
      value: "5",
      metric: "кг",
    },
    info: ["<b>100</b> порций", "<b>5</b> мышей в подарок", "заказчик доволен"],
    isActive: false,
    disabled: {
      value: false,
      warn: "Печалька, с курой закончился.",
    },
    additional: {
      value: "Чего сидишь? Порадуй котэ, <button>купи</button>",
      alternative: "Филе из цыплят с трюфелями в бульоне.",
    },
    fullDescription:'Кошачий корм Нямушка 5 киллограмм, это 100 порций, 5 мышей в подарок'
  },
];
