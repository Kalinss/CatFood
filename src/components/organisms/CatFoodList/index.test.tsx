import React from "react";
import { CatFoodList } from "./index";
import renderer from "react-test-renderer";
import { configure, shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { CatFoodItem } from "../../molecules/CatFoodItem";
import { catFoodData } from "../../../mock/catFoodsDataMock";

configure({ adapter: new Adapter() });

describe("CatFoodList", () => {
  it("should be rendered", () => {
    const tree = renderer
      .create(<CatFoodList data={catFoodData} clickHandler={() => {}} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should contain 3 CatFoodItem components", () => {
    const wrapper = shallow(
      <CatFoodList data={catFoodData} clickHandler={() => {}} />
    );
    expect(wrapper.find(CatFoodItem)).toHaveLength(3);
  });
});
