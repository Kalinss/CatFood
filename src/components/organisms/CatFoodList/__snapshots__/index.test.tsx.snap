// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CatFoodList should be rendered 1`] = `
<div
  className="wrapper"
>
  <ul
    className="list"
  >
    <li>
      <div
        className="wrapper item"
      >
        <div
          className="item"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <svg
            className="border svg"
            id="Слой_2"
            viewBox="0 0 320 480"
            x="0px"
            xmlSpace="preserve"
            y="0px"
          >
            <path
              className="path"
              d="M 305 478 H 12.3 C 6.6 478 2 473.4 2 467.7 V 45 c 0 -3 5 -6 10 -11 c 8.5 -8.5 17.4 -17.5 25 -25 c 5 -5 5.1 -7 8 -7 h 254 c 14.5 -0.5 20.5 -0.5 19 17 v 439 C 317.5 472 318 478 305 478 Z"
            />
          </svg>
          <div
            className="main"
          >
            <p
              className="description mouseLeave"
            >
              Сказочное заморское яство
            </p>
            <div
              className="header title"
            >
               
              <h2>
                Нямушка
              </h2>
               
            </div>
            <div
              className="header subTitle"
            >
               
              <span>
                c фуа-гра
              </span>
               
            </div>
            <div
              className="info"
            >
              <ul>
                <li>
                  <b>
                    10
                  </b>
                   порций
                </li>
                <li>
                  мышь в подарок
                </li>
              </ul>
            </div>
            <div
              className="body price"
            >
              <span
                className="count"
              >
                0,5
              </span>
              <span
                className="metric"
              >
                кг
              </span>
            </div>
          </div>
        </div>
        <p
          className="additional"
          onClick={[Function]}
        >
          Чего сидишь? Порадуй котэ, 
          <a>
            купи
          </a>
        </p>
      </div>
    </li>
    <li>
      <div
        className="wrapper item"
      >
        <div
          className="item"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <svg
            className="border svg"
            id="Слой_2"
            viewBox="0 0 320 480"
            x="0px"
            xmlSpace="preserve"
            y="0px"
          >
            <path
              className="path"
              d="M 305 478 H 12.3 C 6.6 478 2 473.4 2 467.7 V 45 c 0 -3 5 -6 10 -11 c 8.5 -8.5 17.4 -17.5 25 -25 c 5 -5 5.1 -7 8 -7 h 254 c 14.5 -0.5 20.5 -0.5 19 17 v 439 C 317.5 472 318 478 305 478 Z"
            />
          </svg>
          <div
            className="main"
          >
            <p
              className="description mouseLeave"
            >
              Сказочное заморское яство
            </p>
            <div
              className="header title"
            >
               
              <h2>
                Нямушка
              </h2>
               
            </div>
            <div
              className="header subTitle"
            >
               
              <span>
                c рыбой
              </span>
               
            </div>
            <div
              className="info"
            >
              <ul>
                <li>
                  <b>
                    40
                  </b>
                   порций
                </li>
                <li>
                  <b>
                    2
                  </b>
                   мыши в подарок
                </li>
              </ul>
            </div>
            <div
              className="body price"
            >
              <span
                className="count"
              >
                2
              </span>
              <span
                className="metric"
              >
                кг
              </span>
            </div>
          </div>
        </div>
        <p
          className="additional"
          onClick={[Function]}
        >
          Чего сидишь? Порадуй котэ, 
          <a>
            купи
          </a>
        </p>
      </div>
    </li>
    <li>
      <div
        className="wrapper item"
      >
        <div
          className="item"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <svg
            className="border svg"
            id="Слой_2"
            viewBox="0 0 320 480"
            x="0px"
            xmlSpace="preserve"
            y="0px"
          >
            <path
              className="path"
              d="M 305 478 H 12.3 C 6.6 478 2 473.4 2 467.7 V 45 c 0 -3 5 -6 10 -11 c 8.5 -8.5 17.4 -17.5 25 -25 c 5 -5 5.1 -7 8 -7 h 254 c 14.5 -0.5 20.5 -0.5 19 17 v 439 C 317.5 472 318 478 305 478 Z"
            />
          </svg>
          <div
            className="main"
          >
            <p
              className="description mouseLeave"
            >
              Сказочное заморское яство
            </p>
            <div
              className="header title"
            >
               
              <h2>
                Нямушка
              </h2>
               
            </div>
            <div
              className="header subTitle"
            >
               
              <span>
                c курой
              </span>
               
            </div>
            <div
              className="info"
            >
              <ul>
                <li>
                  <b>
                    100
                  </b>
                   порций
                </li>
                <li>
                  <b>
                    5
                  </b>
                   мышей в подарок
                </li>
                <li>
                  заказчик доволен
                </li>
              </ul>
            </div>
            <div
              className="body price"
            >
              <span
                className="count"
              >
                5
              </span>
              <span
                className="metric"
              >
                кг
              </span>
            </div>
          </div>
        </div>
        <p
          className="additional"
          onClick={[Function]}
        >
          Чего сидишь? Порадуй котэ, 
          <a>
            купи
          </a>
        </p>
      </div>
    </li>
  </ul>
</div>
`;
